### ISSUE TITLE
lr_scheduler needs to be at the epoch level

### ISSUE THREAD

**Issue by guidothekp** [huggingface/transformers#35949]
### System Info

In the page: https://huggingface.co/docs/transformers/training, 

note the following blurb.

Isn't lr_scheduler updated per epoch? The code below calls lr_scheduler.step() per batch in stead of per epoch. So the line needs to be moved out to the outer for loop.

```
from tqdm.auto import tqdm

progress_bar = tqdm(range(num_training_steps))

model.train()
for epoch in range(num_epochs):
    for batch in train_dataloader:
        batch = {k: v.to(device) for k, v in batch.items()}
        outputs = model(**batch)
        loss = outputs.loss
        loss.backward()

        optimizer.step()
        lr_scheduler.step()
        optimizer.zero_grad()
        progress_bar.update(1)
```

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Follow the steps in the page above.

### Expected behavior

```
from tqdm.auto import tqdm

progress_bar = tqdm(range(num_training_steps))

model.train()
for epoch in range(num_epochs):
    for batch in train_dataloader:
        batch = {k: v.to(device) for k, v in batch.items()}
        outputs = model(**batch)
        loss = outputs.loss
        loss.backward()

        optimizer.step()
        optimizer.zero_grad()
        progress_bar.update(1)
    lr_scheduler.step()
```

**Comment by Ryoo72** [comment]
Learning rate scheduler updates can be either per-epoch or per-batch depending on configuration. However, modern large models often train on massive datasets with very few epochs (typically 1-3 epochs). In such cases with limited epochs, epoch-based scheduling can be problematic as it provides very few adjustment points for the learning rate. To address this, many modern training approaches calculate the total number of training steps upfront and update the learning rate based on global steps instead of epochs. This is why you'll often see lr_scheduler.step() called within the batch loop rather than the epoch loop in modern training code.
